File: csa_dif.m
function [csax] = csa_dif(x,p,q)
%Generate long memory by using the MA representation of the cross-sectional
%aggregated process using the fast Fourier algorithm, Vera-Valdes(2019)

iT = size(x,1);
n = 2.^nextpow2(2*iT-1);
coefs = ( beta(p+(0:iT-1),q) ./ beta(p,q) ).^(1/2);
csax = ifft(fft(x, n).*fft(coefs', n)); 
csax = csax(1:T, :);
end

File: csa_dif.R
csa_dif = function(x, p, q){
  iT = length(x)
  n = nextn(2*iT - 1, 2)
  k = 0:(iT-1)
  b = (beta(p+k,q)/beta(p,q))^(1/2)
  dx = fft( fft(c(x, rep(0, n - iT))) * fft(c(b, rep(0, n - iT))), inverse = T) / n
  return(Re(dx[1:iT]))
}